Kademlia DHT support (BEP 46 precursor)

garbear opened the pull request:
## Description

This PR adds support for the mainline [Kademlia DHT](https://en.wikipedia.org/wiki/Kademlia). An initial use case of magnet links is provided, allowing media to be browsed and streamed directly from Kademlia.

Magnet links are implemented at the VFS level via the `magnet://` protocol. Magnet links can be added as video sources, browsed, and any files in the magnet link can be played.

BEP 46 is mutable torrents, which would let Kodi store arbitrary data on the DHT, including encrypted userdata.

## Remaining work

~The CMake scripts need improving, which I'm hoping @fuzzard can help with. Boost is especially hacky, and causes the build to break on iOS and tvOS.~ Done!

No disk storage is currently used, everything resides in memory and has to be redownloaded when Kodi is restarted. Clearly some persistence is desired, but I'm surprised how well it works with no persistence whatsoever.

## Related PRs

Requires:

* https://github.com/arvidn/try_signal/pull/21

## Motivation and context

I ultimately want to use Kademlia for game messaging and savestates, so I'm starting with basic DHT read functionality to prove out the dependency.

## How has this been tested?

### Test builds

Included in RetroPlayer test builds: https://github.com/garbear/xbmc/releases

### Runtime testing

Tested on Windows 10 x64 and macOS Ventua 13.5 (M2).

### Sample magnet links

Tested with the Big Buck Bunny magnet link:

```
magnet:?xt=urn:btih:dd8255ecdc7ca55fb0bbf81323d87062db1f6d1c&dn=Big%20Buck%20Bunny
```

## What is the effect on users?

* Added support for magnet links

## Screenshots (if appropriate):

Adding the magnet link as a video source:

![screenshot00001](https://github.com/xbmc/xbmc/assets/531482/a97a61a3-b59e-44c1-adf9-7f781a8690f0)

![screenshot00002](https://github.com/xbmc/xbmc/assets/531482/ad45e2bc-0121-4729-bc42-ae6d3e8174b4)

Browsing the magnet link (thumb generation works):

![screenshot00003](https://github.com/xbmc/xbmc/assets/531482/dfc055c4-cd2e-4dfe-a69a-0a9d01834498)

Playback:

![screenshot00004](https://github.com/xbmc/xbmc/assets/531482/1dfb71b7-201b-458a-abe3-396dee3df3e1)

External subtitles in the magnet link are successfully found.

Kademlia settings:

![screenshot00005](https://github.com/xbmc/xbmc/assets/531482/758148a2-f46a-4ab8-b627-daf166cc1e99)

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

garbear committed:
tools/depends: Add libdatachannel and dependencies

This adds libdatachannel and the following dependencies:

  * libjuice
  * libsrtp
  * plog
  * usrsctp

TODO: Remove the following hacky patches

  * libdatachannel/0001-Look-for-usrsctp-in-subdirectory.patch
  * libdatachannel/0002-Rename-static-library-and-exclude-shared-library.patch
  * usrsctp/0001-Add-CMake-export-targets.patch
  * usrsctp/0002-Also-install-library-to-subdirectory.patch

garbear committed:
tools/depends: Add libtorrent and dependencies

This adds libtorrent and the following dependencies:

  * Boost
  * try_signal

garbear committed:
Add support for magnet URIs

enen92 commented:
Wouldn't it make more sense to create an input stream add-on without adding so many new dependencies to the core itself?

garbear commented:
> I ultimately want to use Kademlia for game messaging, so I'm starting with basic DHT read functionality to prove out the dependency.

Input stream add-on doesn't do any good for game messaging.

I'm planning on encoding my library with HLS and encrypting with AES-128, and using InputStream Adaptive to play back the HLS stream from a magnet link, so an input stream add-on wouldn't work here.

We could do a VFS add-on, but that wouldn't expose the DHT manipulation functions that I need for games.

enen92 commented:
I was just thinking about possible ways to add support for/extending retroplayer without forcing what might be considered "bloat" for everyone else.
What about trying to extend the service addon interface to create some sort of binary service addons? I can see how the project could benefit of such type of addons (e.g. cleaning up the airplay support - now quite dated - and move it to a service instead).

garbear commented:
To quote an aphorism from @notspiff , hero amongst heroes:

```
Once upon time, like 10 years ago i was pondering how to move upnp to an addon and i gave up as i basically need access to the kitchen sink so i get you.
```

Same with Kademlia, its "meat" is the "kitchen sink" we get along with the lib, so any binary add-on work would be massive to support the limited use cases of all p2p functions.

The biggest objection I expect is "dependency bloat - use an add-on!" but we learned from peripheral.joystick that the solution isn't to make mandatory code optional to decrease dependency load - then nothing works.

The solution is dependency management. When you have complexity, manage it. If we want p2p, Kademlia provides the dependencies we need.

joseluismarti commented:
Two questions about Android:

Android dockers are preset to build using ndk 21e, currently do not take the build step `DEFAULT_NDK_VERSION="25c"` into account.

If `use_ndk_api=24` is set, kodi will only run on Android 7.0 or later devices.

garbear commented:
> Android dockers are preset to build using ndk 21e, currently do not take the build step `DEFAULT_NDK_VERSION="25c"` into account.

> If `use_ndk_api=24` is set, kodi will only run on Android 7.0 or later devices.

So this looks like a v22 "P" change, as IP addresses are only exposed in NDK 24+. Unless we implement `get_addrs()` in NDK < 24, Kademlia is Androlid 7.0+ only.

fuzzard commented on CMakeLists.txt:
```diff
@@ -62,8 +62,17 @@ option(ENABLE_AIRTUNES    "Enable AirTunes support?" ON)
+option(ENABLE_LIBTORRENT  "Enable libtorrent support?" ON)
 
 # Internal Depends - supported on all platforms
+option(ENABLE_INTERNAL_BOOST "Enable internal boost?" ON) # TODO: Change to OFF
+option(ENABLE_INTERNAL_LIBDATACHANNEL "Enable internal libdatachannel?" ON) # TODO: Change to OFF
+option(ENABLE_INTERNAL_LIBJUICE "Enable internal libjuice?" ON) # TODO: Change to OFF
+option(ENABLE_INTERNAL_LIBSRTP "Enable internal libsrtp?" ON) # TODO: Change to OFF
+option(ENABLE_INTERNAL_LIBTORRENT "Enable internal libtorrent?" ON) # TODO: Change to OFF
+option(ENABLE_INTERNAL_PLOG "Enable internal plog?" ON) # TODO: Change to OFF
+option(ENABLE_INTERNAL_TRY_SIGNAL "Enable internal try_signal?" ON) # TODO: Change to OFF
+option(ENABLE_INTERNAL_USRSCTP "Enable internal usrsctp?" ON) # TODO: Change to OFF
```
In general you wont need any of this unless its explicitly needed by libkodi. If they are dependencies of eg libtorrent, then the inclusion/search will be handled by the libtorrent find module.

The only exception i can see at a glance is boost, as you actually include a boost header in one of the changes.

garbear replied:
It's complicated, because we actually care about a libtorrent dependency, i.e. libdatachannel, as we want to make WebRTC connections independently.

Originally I had a single option. `ENABLE_INTERNAL_LIBTORRENT`. But then internal libjuice/libsrtp/plog/usrsctp would be dependent on another library, and you mentioned they should have their own "INTERNAL_*" variable instead of cross-depending on libtorrent.

fuzzard replied:
if you care about a dependency, put a REQUIRED on the find call, and handle the failure case.

From what i can see, libtorrent is dependent on libjuice/libsrtp/plog/usrsctp, not the other way around. The enable would be to enable libtorrent, and then in the find module for libtorrent it should handle its dependencies. If those dependences are required, then make their find modules search/or build for that lib. libkodi isnt dependendant on them however.

garbear replied:
If you could look into this later (post-v21, no rush) that would be appreciated. My `magnet-links` branch will remain here for collaboration. And again, post-v21, no rush!

fuzzard commented on CMakeLists.txt:
```diff
@@ -192,6 +205,8 @@ set(required_deps ASS
                   TinyXML
                   TinyXML2
+                  try_signal
+                  Usrsctp
```
Same as in here, only the explicit dependencies used will be in this list, a depencency of a dependency should be included as part of the top dep target (eg libtorrent)

garbear replied:
The problem was that a library-of-a-library wasn't linked into the final kodi.lib on Windows. Adding the dependencies here (even though they're not needed) solved the problem. `Libtorrent` and `LibDataChannel` are the only two dependencies that should be needed in CMakeLists.txt. Can you figure out how to modify the cmake for this?

fuzzard replied:
The target interfaces are wrong then (potentially from what the upstream lib exports), or you are half in/half out with target usage still.

I can, and most likely will, look into it further, but not right now.

fuzzard commented on cmake/modules/FindLibsrtp.cmake:
```diff
@@ -0,0 +1,79 @@
+  else()
+    # Populate paths for find_package_handle_standard_args
+    find_path(LIBSRTP_INCLUDE_DIR NAMES srtp2/srtp.h)
+    find_path(LIBSRTP_LIBRARY NAMES srtp2)
```
This should be find_library presumably.

garbear replied:
Non-internal builds aren't tested, though will need to be if we want libtorrent as a dependency

fuzzard commented on cmake/modules/FindLibsrtp.cmake:
```diff
@@ -0,0 +1,79 @@
+  endif()
+
+  mark_as_advanced(LIBSRTP_INCLUDE_DIR)
+  mark_as_advanced(LIBSRTP_LIBRARY)
```
mark_as_advanced can flag multiple variables in the one call (space delimited)

garbear replied:
I left it as separate calls to better identify variables across all the dependencies I've added. Separate lines mean that I can identify dependencies as they require INCLUDE_DIR, LIBRARY, DEFINITIONS, etc and other variables.

fuzzard replied:
my understanding of mark_as_advanced is it just hides cache variables from cmakegui. Realistically i doubt anyone is actively using that with our project.

garbear replied:
Good point. I think I'd rather leave it as two lines though to improve git history.

fuzzard commented on cmake/modules/FindLibtorrent.cmake:
```diff
@@ -0,0 +1,103 @@
+  else()
+    # Populate paths for find_package_handle_standard_args
+    find_path(LIBTORRENT_INCLUDE_DIR libtorrent/libtorrent.hpp)
+    find_path(LIBTORRENT_LIBRARY NAMES torrent-rasterbar)
```
find_library

fuzzard commented on cmake/modules/FindLibsrtp.cmake:
```diff
@@ -0,0 +1,79 @@
+      LIBSRTP_VERSION)
+
+  if(LIBSRTP_FOUND)
+      set(LIBSRTP_INCLUDE_DIRS "${LIBSRTP_INCLUDE_DIR}")
+      set(LIBSRTP_LIBRARIES "${LIBSRTP_LIBRARY}")
```
If you are going to use the target, dont set these. Will just make my life easier as it gets phased out.

garbear replied:
This will be fixed when me migrate this PR to target usage similar to https://github.com/xbmc/xbmc/pull/23612.

fuzzard commented on cmake/modules/Findtry_signal.cmake:
```diff
@@ -0,0 +1,76 @@
+  else()
+    # Populate paths for find_package_handle_standard_args
+    find_path(TRY_SIGNAL_INCLUDE_DIR NAMES try_signal.hpp)
+    find_path(TRY_SIGNAL_LIBRARY NAMES try_signal)
```
find_library

fuzzard commented on CMakeLists.txt:
```diff
@@ -255,6 +270,10 @@ if(ENABLE_AIRTUNES)
   endif()
 endif()
 
+if(ENABLE_LIBTORRENT)
+  list(APPEND DEP_DEFINES "-DHAS_LIBTORRENT")
+endif()
```
This shouldnt be in here. Going forward the definition should be part of the find module using target_compile_definitions.

https://cmake.org/cmake/help/latest/command/target_compile_definitions.html

Functionally this method will be dependant on https://github.com/xbmc/xbmc/pull/23612

fuzzard commented on tools/depends/target/boost/CMakeLists.txt:
```diff
@@ -0,0 +1,129 @@
+cmake_minimum_required(VERSION 3.16)
```
Should you maybe look at leveraging https://github.com/boostorg/cmake rather than a rag tag custom cmake that potentially becomes a maintenance burden long term.

garbear replied:
Yes, this resource wasn't available 4 years ago when I first ported boost for IPFS/libp2p.

fuzzard commented on tools/depends/target/libsrtp/0002-Enable-OpenSSL.patch:
```diff
@@ -0,0 +1,25 @@
+ set(ERR_REPORTING_STDOUT OFF CACHE BOOL "Enable logging to stdout")
+ set(ERR_REPORTING_FILE "" CACHE FILEPATH "Use file for logging")
+-set(ENABLE_OPENSSL OFF CACHE BOOL "Enable OpenSSL crypto engine")
++set(ENABLE_OPENSSL ON CACHE BOOL "Enable OpenSSL crypto engine")
```
These should be set as arguments rather than patching.

https://cmake.org/cmake/help/latest/command/set.html

```
Sets the given cache <variable> (cache entry). Since cache entries are meant to provide user-settable values this does not overwrite existing cache entries by default. Use the FORCE option to overwrite existing entries.
```

the cmake build argument should just have -DENABLE_OPENSSL=ON

garbear commented:
@fuzzard Thanks for your cmake reviews, the suggestions are most valuable.

I put so much working into making libtorrent "just barely work" on so many platforms, in order to get Jenkins partially passing for this PR. Now that the PR is open, I would appreciate as many cmake fixes to my `magnet-links` branch as your robot body can pump out.

fuzzard commented:
Yeah, i'll have some spare time come christmas, so if its good for v22, i can dig into it then. Hopefully i'll have other cmake stuff merged to make it a bit easier

garbear commented:
Yeah, I think shipping Kademlia with v22 is doable, if just barely.

By then our use case won't be a single magnet link. We can even artificially deny VideoPlayer to play back magnet links for the optics. I hope to be using BEP 46, mutable torrents, to start sharing Kodi userdata encrypted on the DHT between peers.

garbear commented:
@fuzzard I addressed all comments not related to TARGET usage or redoing Boost with its cmake scripts.

I had some difficulties cross-compiling libdatachannel that left some hacky patches in the code, like renaming a library so my (faulty) cmake would find them. They are documented in the libdatachannel commit.

If you could get the CMake working later for v22 that'd be appreciated.

garbear commented:
Jenkins is now passing on all platforms except for the following:

* iOS/tvOS - Needs Boost updated to its CMake scripts. Error is:

```
fileunix.cpp:103:11: fatal error: 'ar.h' file not found
# include <ar.h>
          ^~~~~~
1 error generated.

Failed to build B2 build engine
```

* Windows UWP 64/ARM (but not 32). Error is:

```
WindowsApp.lib(api-ms-win-core-memory-l1-1-0.dll) : error LNK2005: CreateFileMappingW already defined in avutil.lib(file.o) [kodi.vcxproj]
WindowsApp.lib(api-ms-win-core-memory-l1-1-0.dll) : error LNK2005: MapViewOfFile already defined in avutil.lib(file.o) [kodi.vcxproj]

kodi.exe : fatal error LNK1169: one or more multiply defined symbols found [kodi.vcxproj]
```

garbear commented:
Rebased on v21 Beta 1

da-anda commented:
> do we want to be forever confined to a client-server model
wouldn't we need a proper client/server implementation first in order to be confined to it? All we currently have is hack-ish centralized/shared DB, but no client/server functionality in it's actual meaning (like Plex, Emby, Jellyfin) ;) 

Also, will people understand what a "Kademlia" setting means or should we abstract the low level implementation with a naming that average people might be more familiar with?

garbear commented:
UPnP is client/server. VFS is client, http is server. Client/server shows up everywhere, but our codebase has no p2p.

Good idea, let's call the setting "Pirate hacky magnet thing" 😉

The motivation was p2p messaging using Kademlia. The fact that it can play back magnet links is just a side effect. I went with Kademlia because it distances us from unintentional piracy use cases.

garbear commented:
jenkins build this please

garbear committed:
[UWP] Enable strict C++20 coroutine support

Fixes error with dependencies that use <coroutine>:

  coroutine(15): warning STL4039: The contents of <coroutine> are not available
  with /await. Remove /await or use /await:strict for standard coroutines. Use
  <experimental/coroutine> for legacy /await support.

garbear commented:
I've rebased on master and fixed all platforms except for two with Boost assembly errors (Linux arm32 and UWP arm). A future version of Boost may be needed to fix these.

zag2me commented:
Will this work with a playlist or some kind of shortcut? 
I'd love to test but typing in that big string makes it very difficult!

garbear commented:
> Will this work with a playlist or some kind of shortcut?

It should. It makes `magenet:` URIs valid URIs in kodi. But while the VFS should work perfectly, there's still some stuff missing, like persistent storage.

zag2me commented:
Tested with your build KodiSetup-20250214-7a63b0d0-retroplayer-21.2-x64 but unfortunately didn't work for me.

Files >> Add Video >> click on none >> copy and paste magnet link >> Rename name to Big Buck Bunny. Hit play.

I got a warning about a network error then unable to play item. "Remote Share - Path not found or invalid".

https://paste.kodi.tv/iwiwanazaf.kodi

garbear commented:
Hm, this was dropped from RetroPlayer test builds, you'll see the Kademlia PR isn't in the 20250214 release notes. You can go back far enough before I dropped this from test builds, or build this branch. Note that you'll need to enable "Kademlia" in settings as it's default disabled due to pinging anonymous peers when enabled.

garbear commented:
> I'm also not sure on the need for persistence, personally I just want to stream stuff these days and input or host a "source". Needs to be simple.

Interesting, I guess persistence isn't needed if your model is just to stream the stuff you're watching. Restarting the stream /or kodi will clear the cache, but I guess that doesn't matter if you don't re-watch part of the video.

> Would love to be able to add a HTTPS source for a folder or something like that of magnet links and be able to use Kodi library on it with scrapers.

Cool use case. This should probably work already, if you wanna give it a try. And here's an idea: Use a magnet link to a folder with other magnet links.

> Overall I think this is great and probably the thing to revive Kodi from its current irrelevance in the days of JellyFin/Emby and Streemio dominance.

Agreed. However the team feels direct torrent support would cause more people and automated systems to mis-associate us with piracy. If we could use Kademlia for messaging, then torrent support would just be a side effect that we'd allow. And opinions can change.

garbear commented on xbmc/network/torrent/TorrentUtils.cpp:
```diff
@@ -0,0 +1,30 @@
+
+  // Remove any protocol slashes from the magnet URI
+  if (StringUtils::StartsWith(magnetUri, "magnet://"))
+    magnetUri.erase(std::strlen("magnet:"), 2);
```
Incorrect I think

garbear commented:
@zag2me I fixed WIN-ARM64, so in theory it could go in. But I think we should bring up Kademlia support at next devcon and see if views have changed.

acidzab commented:
@garbear @zag2me curious about that feature, how could this can be applied in a "music server" scenario with Kodi ? Theoretically music server is a wrong scenario because with dht each kodi connects together accessing various music libraries

garbear commented:
Jenkins build this please

garbear commented:
Jenkins build this please
